### PR TITLE
Refactor terminal - a first attempt

### PR DESCRIPTION
**Pull request by sherter**
Please focus on the differences in the `Terminal` and `Token` class and tell me what you think. I didn't fix all the tests just yet, in case you don't like the proposal.

### CODE REVIEW

File: src/main/java/minijava/token/Terminal.java [hunk 1]
```diff
@@ -1,114 +1,92 @@
...
+    LEFT,
+    RIGHT
   }
 
-  public TerminalType getType() {
-    return terminalType;
-  }
+  public final Optional<String> string;
+  final Optional<Associativity> associativity;
+  final Optional<Integer> precedence;
```
Comment (sgraf812) [inline src/main/java/minijava/token/Terminal.java]: Well, the public field is OK I guess, but I found this rather interesting: http://blog.joda.org/2015/08/java-se-8-optional-pragmatic-approach.html

TLDR; Use `Optional` only as return value from public APIs, otherwise use `null` (and document so).

**Review by sgraf812 — APPROVED** [review]
I like where this is going

File: src/main/java/minijava/token/Terminal.java (hunk 1 above)
Comment (sherter) [reply src/main/java/minijava/token/Terminal.java]: Looks much better now :+1:

File: src/main/java/minijava/token/Token.java [hunk 2]
```diff
@@ -1,50 +1,59 @@
...
+  public boolean isOperator() {
+    return terminal.associativity != null;
   }
 
-  public boolean isType(Terminal.TerminalType type) {
-    return terminal.isType(type);
-  }
-
-  @Override
-  public String toString() {
```
Comment (parttimenerd) [inline src/main/java/minijava/token/Token.java:29]: It might be useful to have a function in the refactored Token class that outputs the same information for error messages.

**Comment by parttimenerd** [comment]
I like the current refactorings too.

**Comment by sgraf812** [comment]
I could help with updating `SimpleLexerProperties` (missed to rename it to `LexerProperties`), but I haven't really time for the rest of the weekend.

**Comment by sherter** [comment]
I'm working on it (it's a little tricky though). Maybe I'll remind you of that offer, if I can't figure it out :wink:

**Comment by sherter** [comment]
Whitespace and comments are currently not inserted randomly, because the corresponding tokens were removed. Otherwise I think this is ready for merging.

File: src/main/java/minijava/token/Token.java (hunk 2 above)
Comment (sherter) [reply src/main/java/minijava/token/Token.java:29]: I'm not sure what you mean by that. Do you need a method that outputs something different from the `toString` method?

Comment (parttimenerd) [reply src/main/java/minijava/token/Token.java:29]: Yes, to simplify the creation of error messages.
But it doesn't matter in this context (we can just change der ParserError class).
